### ISSUE TITLE
[flow] Parsing error when calling generic functions with type arguments when flow pragma is not first comment 

### ISSUE THREAD

**Issue by FermiDirak** [babel/babel#9240]
## Bug Report

**Current Behavior**
When the @flow pragma is not specified as the first comment of a JS or JSX file, babel parser will not be able to parse any code that passes type arguments.

**Input Code**
```js
// arbitrary comment
// @flow

const n: number = 5; // this works fine

const set = new Set<number>([1, 2, 3]);
        // unexpected token^
```

**Expected behavior/code**
Flow should be able to parse functions that are called with type arguments so long as the @flow pragma exists before any code. 

**Babel Configuration (.babelrc)**

```js
{
  presets: [
    "@babel/preset-env",
    "@babel/preset-react",
    "@babel/preset-flow",
  ]
}
```

**Environment**
- Babel version: [^7.0.0]
- Node/npm version: [Node 8.11.1]
- OS: [OSX 10.13.4]
- Monorepo: [yes]
- How you are using Babel: [`loader`]

**Possible Solution**

It appears this line ignores the flow pragma if it isn't found in the first comment of the source file.
https://github.com/babel/babel/blob/c586d4e8cac816941bf2ba559472fd9cc3150921/packages/babel-parser/src/plugins/flow.js#L77

This could be changed a-la Flow-Strip-Types, which first checks all comments for a flow pragma before it begins transforming. (https://github.com/babel/babel/blob/c586d4e8cac816941bf2ba559472fd9cc3150921/packages/babel-plugin-transform-flow-strip-types/src/index.js). Perhaps a solution like this could be used here as well.

**Comment by nateshmbhat** [comment]
I had this problem because of older babel version which I fixed using a stackoverflow answer  : 
Check this stackoverflow answer : https://stackoverflow.com/a/53927430/6665568

**Comment by JakubRimal** [comment]
@FermiDirak Thank you very much, you saved me at least a few hours as without seeing this issue I would never guess that the problem is because of a position of the `@flow`... :open_mouth: :smiley:

**Comment by pwlmaciejewski** [comment]
I'm experiencing the same issue with flow enabled by default on all files:

```
// .flowconfig
[options]
all=true
```

Adding `@flow` pragma in the first line of the throwing files fixes fixes it.

**Comment by nicolo-ribaudo** [comment]
Babel doesn't read your `.flowconfig` file.

**Comment by pwlmaciejewski** [comment]
@nicolo-ribaudo yes, but afaik `babel-preset-flow` doesn't expose any options, or at least they are not documented here: https://babeljs.io/docs/en/babel-preset-flow. 

It comes with `babel-plugin-transform-flow-strip-types` that has `requireDirective` option set to `false` by default. If I get it correctly, it should strip the types out of the box. Or maybe `babel-plugin-transform-flow-strip-types` has nothing to do with this error?

**Comment by nicolo-ribaudo** [comment]
That preset taks an `"all": boolean` option (and also the plugin). It's a documentation bug.

**Comment by existentialism** [comment]
Ref: https://github.com/babel/website/issues/1973

**Comment by pwlmaciejewski** [comment]
Thanks, @nicolo-ribaudo, it works :+1: 

In my case it was `babel-preset-expo` that uses `flow-strip-types`, but this one does the trick:

```
overrides: [{
    plugins: [
        ['@babel/plugin-transform-flow-strip-types', {all: true}],
    ]
}]
```

**Comment by loganfsmyth** [comment]
Looks like this is probably because of https://github.com/babel/babel/blob/165ef2941694184d83307b41d311751bd3dabae0/packages/babel-parser/src/plugins/flow.js#L103 which will set `flowPragma` to `null` for the first comment it sees, meaning that it will never try to process later comments. 

That line should just be removed since the logic in https://github.com/babel/babel/blob/165ef2941694184d83307b41d311751bd3dabae0/packages/babel-parser/src/plugins/flow.js#L1601 is supposed to handle that. That said, I'm not sure this `parseStatement` call is actually a good place for it, or at least we want to make sure that code like

```
(/* @flow */ 42);
```
also doesn't count (I assume), but that comment isn't after any statement has finished.

**Comment by julienw** [comment]
Just a quick note that this affects prettier as well, because prettier uses babel to parse JS files by default. Setting prettier's parser to `babel-flow` works but has issues as well when we want to use prettier for different file types (eg JS and CSS).

**Comment by tanhauhau** [comment]
after some investigation on the flow's source code, i noticed that the logic babel used to determine whether a file is deemed a "flow" file is different than [flow's implementation](https://github.com/facebook/flow/blob/master/src/parsing/parsing_service_js.ml#L275)

Basically, in flow, it 
- check [`docblock_max_tokens`](https://github.com/facebook/flow/blob/master/src/parsing/parsing_service_js.ml#L141) tokens, [which is default to `10`](https://flow.org/en/docs/config/options/#toc-max-header-tokens-integer), to look for any sign of comments that matches `@flow` comment
  - if it encounter any other token than `CommentToken`, [it will early exit](https://github.com/facebook/flow/blob/master/src/parsing/parsing_service_js.ml#L301)
    - with the exception of strings and semicolons, which they assume would be `"use strict";`

in babel-parser flow-plugin, we look for
- first comment
  - if it is a flow comment, deemed as flow file
  - if it is not, deemed as not a flow file

one edge case for this logic would be
```js
a<x>(y);
// @flow
a<x>(y);
```
[babel repl](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=IYHgHgfAFAnglAbgFAHoUAIACAzANgewHclRJZEg&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Cflow%2Cenv&prettier=false&targets=&version=7.4.3&externalPlugins=%40babel%2Fplugin-syntax-flow%407.2.0) 

anything before the first flow comment would be parsed as non-flow js, and anything after it would be treated as flow, eg: first statement treated as `BinaryExpression`, and second statement as `CallExpression` with `TypeParameterInstantiation`.

**Comment by tanhauhau** [comment]
I would like to contribute to fix this behavior, by adding a similar `max_header_tokens` options to `flow syntax`, just like what we did with `all`

I've started with some test cases and boilerplates, https://github.com/babel/babel/pull/9885, but I am not sure on how to lookahead and parse the first `n` tokens,

**Comment by nicolo-ribaudo** [comment]
Fixed by #9891
